src/libcore/iter/adapters/mod.rs

```diff
@@ -707,12 +702,9 @@ impl<I: Iterator, P> Iterator for Filter<I, P> where P: FnMut(&I::Item) -> bool
     // Using the branchless version will also simplify the LLVM byte code, thus
     // leaving more budget for LLVM optimizations.
     #[inline]
-    fn count(mut self) -> usize {
-        let mut count = 0;
-        for x in &mut self.iter {
-            count += (self.predicate)(&x) as usize;
-        }
-        count
+    fn count(self) -> usize {
+        let mut predicate = self.predicate;
+        self.iter.map(|x| predicate(&x) as usize).sum()
```

Comment:
Since there's a big perf comment above this, I think this should include a benchmark and some results from you showing that it's no worse for simple cases, and better in some.  Maybe something like the ones in https://github.com/rust-lang/rust/blob/master/src/libcore/benches/iter.rs#L161